File: tm-go/parsers/js/lexer.go
package js

import (
	"bytes"
	"unicode/utf8"
)

const (
	State_initial = 0
	State_div = 1
	State_template = 2
	State_template_div = 3
)

// ErrorHandler is called every time a lexer or parser is unable to process
// some part of the input.
type ErrorHandler func(line, offset, len int, msg string)

// Lexer uses a generated DFA to scan through a utf-8 encoded input string. If
// the string starts with a BOM character, it gets skipped.
type Lexer struct {
	source []byte
	err    ErrorHandler

	ch          rune // current character, -1 means EOI
	offset      int  // character offset
	tokenOffset int  // last token offset
	line        int  // current line number (1-based)
	tokenLine   int  // last token line
	lineOffset  int  // current line offset
	scanOffset  int  // scanning offset
	value       interface{}

	State int // lexer state, modifiable
}

const bom = 0xfeff // byte order mark, permitted as a first character only
var bomSeq = []byte{0xEF, 0xBB, 0xBF}

// Init prepares the lexer l to tokenize source by performing the full reset
// of the internal state.
//
// Note that Init may call err one or more times if there are errors in the
// first few characters of the text.
func (l *Lexer) Init(source []byte, err ErrorHandler) {
	l.source = source
	l.err = err

	l.ch = 0
	l.offset = 0
	l.tokenOffset = 0
	l.line = 1
	l.tokenLine = 1
	l.lineOffset = 0
	l.scanOffset = 0
	l.State = 0

	if bytes.HasPrefix(source, bomSeq) {
		l.scanOffset += len(bomSeq)
	}

skipChar:
	l.offset = l.scanOffset
	if l.offset < len(l.source) {
		r, w := rune(l.source[l.offset]), 1
		if r >= 0x80 {
			// not ASCII
			r, w = utf8.DecodeRune(l.source[l.offset:])
			if r == utf8.RuneError && w == 1 || r == bom {
				l.invalidRune(r, w)
				l.scanOffset += w
				goto skipChar
			}
		}
		l.scanOffset += w
		l.ch = r
	} else {
		l.ch = -1 // EOI
	}
}

// Next finds and returns the next token in l.source. The source end is
// indicated by Token.EOI.
//
// The token text can be retrieved later by calling the Text() method.
func (l *Lexer) Next() Token {
	prevLine := l.tokenLine
restart:
	l.tokenLine = l.line
	l.tokenOffset = l.offset

	state := tmStateMap[l.State]
    hash := uint32(0)
	for state >= 0 {
		var ch int
		switch {
		case l.ch < 0:
			ch = 0
		case int(l.ch) < len(tmRuneClass):
			ch = int(tmRuneClass[l.ch])
		default:
			ch = mapRune(l.ch)
		}
		state = int(tmLexerAction[state*tmNumClasses+ch])
		if state == -1 && ch == -1 {
			l.err(l.line, l.tokenOffset, l.offset-l.tokenOffset, "Unexpected end of input reached")
			return EOI
		}
		if state < -1 || ch == -1 {
			break
		}
    	hash = hash*uint32(31) + uint32(l.ch)

		if l.ch == '\n' {
			l.line++
			l.lineOffset = l.offset
		}
	skipChar:
		// Scan the next character.
		// Note: the following code is inlined to avoid performance implications.
		l.offset = l.scanOffset
		if l.offset < len(l.source) {
			r, w := rune(l.source[l.offset]), 1
			if r >= 0x80 {
				// not ASCII
				r, w = utf8.DecodeRune(l.source[l.offset:])
				if r == utf8.RuneError && w == 1 || r == bom {
					l.invalidRune(r, w)
					l.scanOffset += w
					goto skipChar
				}
			}
			l.scanOffset += w
			l.ch = r
		} else {
			l.ch = -1 // EOI
		}
	}
	if state == -1 {
		l.err(l.tokenLine, l.tokenOffset, l.offset-l.tokenOffset, "invalid token")
		goto restart
	}
	if state == -2 {
		return EOI
	}

	rule := -state - 3
	switch rule {
	case 0:
		hh := hash&63
		switch hh {
		case 1:
			if hash == 0x5c13d641 && bytes.Equal([]byte("default"), l.source[l.tokenOffset:l.offset]) {
				rule = 13
				break
			}
			if hash == 0x2f9501 && bytes.Equal([]byte("enum"), l.source[l.tokenOffset:l.offset]) {
				rule = 40
				break
			}
		case 3:
			if hash == 0xcd244983 && bytes.Equal([]byte("finally"), l.source[l.tokenOffset:l.offset]) {
				rule = 19
				break
			}
		case 6:
			if hash == 0x37b0c6 && bytes.Equal([]byte("with"), l.source[l.tokenOffset:l.offset]) {
				rule = 37
				break
			}
		case 7:
			if hash == 0x33c587 && bytes.Equal([]byte("null"), l.source[l.tokenOffset:l.offset]) {
				rule = 41
				break
			}
		case 9:
			if hash == 0x18cc9 && bytes.Equal([]byte("for"), l.source[l.tokenOffset:l.offset]) {
				rule = 20
				break
			}
		case 11:
			if hash == 0xc8b && bytes.Equal([]byte("do"), l.source[l.tokenOffset:l.offset]) {
				rule = 15
				break
			}
		case 13:
			if hash == 0x6da5f8d && bytes.Equal([]byte("yield"), l.source[l.tokenOffset:l.offset]) {
				rule = 38
				break
			}
		case 14:
			if hash == 0x36758e && bytes.Equal([]byte("true"), l.source[l.tokenOffset:l.offset]) {
				rule = 42
				break
			}
		case 17:
			if hash == 0xcb7e7191 && bytes.Equal([]byte("target"), l.source[l.tokenOffset:l.offset]) {
				rule = 51
				break
			}
			if hash == 0xcccfb691 && bytes.Equal([]byte("typeof"), l.source[l.tokenOffset:l.offset]) {
				rule = 33
				break
			}
		case 20:
			if hash == 0x375194 && bytes.Equal([]byte("void"), l.source[l.tokenOffset:l.offset]) {
				rule = 35
				break
			}
		case 22:
			if hash == 0x58e7956 && bytes.Equal([]byte("await"), l.source[l.tokenOffset:l.offset]) {
				rule = 39
				break
			}
			if hash == 0x18f56 && bytes.Equal([]byte("get"), l.source[l.tokenOffset:l.offset]) {
				rule = 46
				break
			}
		case 23:
			if hash == 0xdd7 && bytes.Equal([]byte("of"), l.source[l.tokenOffset:l.offset]) {
				rule = 48
				break
			}
		case 24:
			if hash == 0x524f73d8 && bytes.Equal([]byte("function"), l.source[l.tokenOffset:l.offset]) {
				rule = 21
				break
			}
		case 25:
			if hash == 0xb22d2499 && bytes.Equal([]byte("extends"), l.source[l.tokenOffset:l.offset]) {
				rule = 18
				break
			}
		case 27:
			if hash == 0x1a21b && bytes.Equal([]byte("let"), l.source[l.tokenOffset:l.offset]) {
				rule = 47
				break
			}
		case 29:
			if hash == 0xd1d && bytes.Equal([]byte("if"), l.source[l.tokenOffset:l.offset]) {
				rule = 22
				break
			}
		case 30:
			if hash == 0x364e9e && bytes.Equal([]byte("this"), l.source[l.tokenOffset:l.offset]) {
				rule = 30
				break
			}
		case 32:
			if hash == 0x1a9a0 && bytes.Equal([]byte("new"), l.source[l.tokenOffset:l.offset]) {
				rule = 26
				break
			}
		case 33:
			if hash == 0x20a6f421 && bytes.Equal([]byte("debugger"), l.source[l.tokenOffset:l.offset]) {
				rule = 12
				break
			}
		case 34:
			if hash == 0x1bc62 && bytes.Equal([]byte("set"), l.source[l.tokenOffset:l.offset]) {
				rule = 49
				break
			}
		case 35:
			if hash == 0x5a73763 && bytes.Equal([]byte("const"), l.source[l.tokenOffset:l.offset]) {
				rule = 10
				break
			}
			if hash == 0x5cb1923 && bytes.Equal([]byte("false"), l.source[l.tokenOffset:l.offset]) {
				rule = 43
				break
			}
		case 37:
			if hash == 0xb96173a5 && bytes.Equal([]byte("import"), l.source[l.tokenOffset:l.offset]) {
				rule = 23
				break
			}
			if hash == 0xd25 && bytes.Equal([]byte("in"), l.source[l.tokenOffset:l.offset]) {
				rule = 24
				break
			}
		case 38:
			if hash == 0x693a6e6 && bytes.Equal([]byte("throw"), l.source[l.tokenOffset:l.offset]) {
				rule = 31
				break
			}
		case 39:
			if hash == 0xde312ca7 && bytes.Equal([]byte("continue"), l.source[l.tokenOffset:l.offset]) {
				rule = 11
				break
			}
			if hash == 0x1c727 && bytes.Equal([]byte("var"), l.source[l.tokenOffset:l.offset]) {
				rule = 34
				break
			}
		case 42:
			if hash == 0x3017aa && bytes.Equal([]byte("from"), l.source[l.tokenOffset:l.offset]) {
				rule = 45
				break
			}
		case 43:
			if hash == 0xb06685ab && bytes.Equal([]byte("delete"), l.source[l.tokenOffset:l.offset]) {
				rule = 14
				break
			}
		case 44:
			if hash == 0x35c3d12c && bytes.Equal([]byte("instanceof"), l.source[l.tokenOffset:l.offset]) {
				rule = 25
				break
			}
		case 46:
			if hash == 0xcacdce6e && bytes.Equal([]byte("static"), l.source[l.tokenOffset:l.offset]) {
				rule = 50
				break
			}
		case 48:
			if hash == 0x2e7b30 && bytes.Equal([]byte("case"), l.source[l.tokenOffset:l.offset]) {
				rule = 7
				break
			}
			if hash == 0xc84e3d30 && bytes.Equal([]byte("return"), l.source[l.tokenOffset:l.offset]) {
				rule = 27
				break
			}
		case 49:
			if hash == 0x6bdcb31 && bytes.Equal([]byte("while"), l.source[l.tokenOffset:l.offset]) {
				rule = 36
				break
			}
		case 50:
			if hash == 0xc32 && bytes.Equal([]byte("as"), l.source[l.tokenOffset:l.offset]) {
				rule = 44
				break
			}
		case 52:
			if hash == 0xb32913b4 && bytes.Equal([]byte("export"), l.source[l.tokenOffset:l.offset]) {
				rule = 17
				break
			}
			if hash == 0xcafbb734 && bytes.Equal([]byte("switch"), l.source[l.tokenOffset:l.offset]) {
				rule = 29
				break
			}
		case 56:
			if hash == 0x5a5a978 && bytes.Equal([]byte("class"), l.source[l.tokenOffset:l.offset]) {
				rule = 9
				break
			}
		case 57:
			if hash == 0x2f8d39 && bytes.Equal([]byte("else"), l.source[l.tokenOffset:l.offset]) {
				rule = 16
				break
			}
		case 59:
			if hash == 0x5a0eebb && bytes.Equal([]byte("catch"), l.source[l.tokenOffset:l.offset]) {
				rule = 8
				break
			}
			if hash == 0x68b6f7b && bytes.Equal([]byte("super"), l.source[l.tokenOffset:l.offset]) {
				rule = 28
				break
			}
			if hash == 0x1c1bb && bytes.Equal([]byte("try"), l.source[l.tokenOffset:l.offset]) {
				rule = 32
				break
			}
		case 63:
			if hash == 0x59a58ff && bytes.Equal([]byte("break"), l.source[l.tokenOffset:l.offset]) {
				rule = 6
				break
			}
		}
	}

	token := tmToken[rule]
	space := false
	switch rule {
	case 2: // WhiteSpace: /[\t\v\f \xa0\ufeff\p{Zs}]/
		space = true
	case 3: // LineTerminatorSequence: /[\n\r\u2028\u2029]|\r\n/
		space = true
	case 4: // MultiLineComment: /\/\*{commentChars}\*\//
		space = true
	case 5: // SingleLineComment: /\/\/[^\n\r\u2028\u2029]*/
		space = true
	}
	if space {
		goto restart
	}

	// There is an ambiguity in the language that a slash can either represent
	// a division operator, or start a regular expression literal. This gets
	// disambiguated at the grammar level - division always follows an
	// expression, while regex literals are expressions themselves. Here we use
	// some knowledge about the grammar to decide whether the next token can be
	// a regular expression literal.
	//
	// See the following thread for more details:
	// http://stackoverflow.com/questions/5519596/when-parsing-javascript-what

	inTemplate := l.State >= State_template
	var reContext bool
	switch token {
	case NEW, DELETE, VOID, TYPEOF, INSTANCEOF, IN, DO, RETURN, CASE, THROW, ELSE:
		reContext = true
	case TEMPLATEHEAD, TEMPLATEMIDDLE:
		reContext = true
		inTemplate = true
	case TEMPLATETAIL:
		reContext = false
		inTemplate = false
	case RPAREN, RBRACK:
		// TODO support if (...) /aaaa/;
		reContext = false
	case PLUSPLUS, MINUSMINUS:
		if prevLine != l.tokenLine {
			// This is a pre-increment/decrement, so we expect a regular expression.
			reContext = true
		} else {
			// If we were in reContext before this token, this is a
			// pre-increment/decrement, otherwise, this is a post. We can just
			// propagate the previous value of reContext.
			reContext = l.State == State_template || l.State == State_initial
		}
	default:
		reContext = token >= punctuationStart && token < punctuationEnd
	}
	if inTemplate {
		if reContext {
			l.State = State_template
		} else {
			l.State = State_template_div
		}
	} else if reContext {
		l.State = State_initial
	} else {
		l.State = State_div
	}
	return token
}

func (l *Lexer) invalidRune(r rune, w int) {
	switch r {
	case utf8.RuneError:
		l.err(l.line, l.offset, w, "illegal UTF-8 encoding")
	case bom:
		l.err(l.line, l.offset, w, "illegal byte order mark")
	}
}

// Pos returns the start and end positions of the last token returned by Next().
func (l *Lexer) Pos() (start, end int) {
	start = l.tokenOffset
	end = l.offset
	return
}

// Line returns the line number of the last token returned by Next().
func (l *Lexer) Line() int {
	return l.tokenLine
}

// Text returns the substring of the input corresponding to the last token.
func (l *Lexer) Text() string {
	return string(l.source[l.tokenOffset:l.offset])
}

func (l *Lexer) Value() interface{} {
	return l.value
}

File: examples/cpp/json/lexer.cc
// generated through human effort; PLEASE DO EDIT

#include "lexer.h"

#include "third_party/absl/log/log.h"
#include "third_party/absl/strings/match.h"

namespace json {

namespace {
constexpr int tmNumClasses = 24;

constexpr uint8_t tmRuneClass[] = {
    1,  1,  1,  1,  1,  1,  1,  1,  1,  2,  2,  1,  1,  2,  1,  1,  1,  1,
    1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  1,  2,  1,  3,  1,
    1,  1,  1,  1,  1,  1,  4,  5,  6,  7,  8,  9,  10, 11, 11, 11, 11, 11,
    11, 11, 11, 11, 12, 1,  1,  1,  1,  1,  1,  13, 13, 13, 13, 14, 13, 15,
    15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15, 15,
    15, 16, 17, 18, 1,  1,  1,  13, 19, 13, 13, 14, 19, 15, 15, 15, 15, 15,
    15, 15, 20, 15, 15, 15, 20, 15, 20, 21, 15, 15, 15, 15, 15, 22, 1,  23,
};

constexpr int tmRuneClassLen = 126;
constexpr int tmFirstRule = -3;

constexpr int tmStateMap[] = {
    0,
};

constexpr int8_t tmLexerAction[] = {
    -3,  -4,  27,  20,  -4,  -4,  19,  18,  -4,  14,  13,  7,   6,   5,   5,
    5,   4,   -4,  3,   5,   5,   5,   2,   1,   -6,  -6,  -6,  -6,  -6,  -6,
    -6,  -6,  -6,  -6,  -6,  -6,  -6,  -6,  -6,  -6,  -6,  -6,  -6,  -6,  -6,
    -6,  -6,  -6,  -5,  -5,  -5,  -5,  -5,  -5,  -5,  -5,  -5,  -5,  -5,  -5,
    -5,  -5,  -5,  -5,  -5,  -5,  -5,  -5,  -5,  -5,  -5,  -5,  -8,  -8,  -8,
    -8,  -8,  -8,  -8,  -8,  -8,  -8,  -8,  -8,  -8,  -8,  -8,  -8,  -8,  -8,
    -8,  -8,  -8,  -8,  -8,  -8,  -7,  -7,  -7,  -7,  -7,  -7,  -7,  -7,  -7,
    -7,  -7,  -7,  -7,  -7,  -7,  -7,  -7,  -7,  -7,  -7,  -7,  -7,  -7,  -7,
    -15, -15, -15, -15, -15, -15, -15, -15, -15, -15, 5,   5,   -15, 5,   5,
    5,   -15, -15, -15, 5,   5,   5,   -15, -15, -9,  -9,  -9,  -9,  -9,  -9,
    -9,  -9,  -9,  -9,  -9,  -9,  -9,  -9,  -9,  -9,  -9,  -9,  -9,  -9,  -9,
    -9,  -9,  -9,  -14, -14, -14, -14, -14, -14, -14, -14, -1,  -14, 7,   7,
    -14, -14, -2,  -14, -14, -14, -14, -14, -14, -14, -14, -14, -4,  -4,  -4,
    -4,  -4,  10,  -4,  10,  -4,  -4,  9,   9,   -4,  -4,  -4,  -4,  -4,  -4,
    -4,  -4,  -4,  -4,  -4,  -4,  -14, -14, -14, -14, -14, -14, -14, -14, -14,
    -14, 9,   9,   -14, -14, -14, -14, -14, -14, -14, -14, -14, -14, -14, -14,
    -4,  -4,  -4,  -4,  -4,  -4,  -4,  -4,  -4,  -4,  9,   9,   -4,  -4,  -4,
    -4,  -4,  -4,  -4,  -4,  -4,  -4,  -4,  -4,  -4,  -4,  -4,  -4,  -4,  -4,
    -4,  -4,  -4,  -4,  12,  12,  -4,  -4,  -4,  -4,  -4,  -4,  -4,  -4,  -4,
    -4,  -4,  -4,  -14, -14, -14, -14, -14, -14, -14, -14, -14, -14, 12,  12,
    -14, -14, -2,  -14, -14, -14, -14, -14, -14, -14, -14, -14, -14, -14, -14,
    -14, -14, -14, -14, -14, -1,  -14, -14, -14, -14, -14, -2,  -14, -14, -14,
    -14, -14, -14, -14, -14, -14, -4,  -4,  -4,  -4,  15,  -4,  -4,  -4,  -4,
    -4,  -4,  -4,  -4,  -4,  -4,  -4,  -4,  -4,  -4,  -4,  -4,  -4,  -4,  -4,
    -4,  15,  15,  15,  16,  15,  15,  15,  15,  15,  15,  15,  15,  15,  15,
    15,  15,  15,  15,  15,  15,  15,  15,  15,  -4,  15,  15,  15,  16,  15,
    15,  15,  15,  17,  15,  15,  15,  15,  15,  15,  15,  15,  15,  15,  15,
    15,  15,  15,  -12, -12, -12, -12, -12, -12, -12, -12, -12, -12, -12, -12,
    -12, -12, -12, -12, -12, -12, -12, -12, -12, -12, -12, -12, -4,  -4,  -4,
    -4,  -4,  -4,  -4,  -4,  -4,  -4,  13,  7,   -4,  -4,  -4,  -4,  -4,  -4,
    -4,  -4,  -4,  -4,  -4,  -4,  -10, -10, -10, -10, -10, -10, -10, -10, -10,
    -10, -10, -10, -10, -10, -10, -10, -10, -10, -10, -10, -10, -10, -10, -10,
    -4,  20,  20,  26,  20,  20,  20,  20,  20,  20,  20,  20,  20,  20,  20,
    20,  20,  21,  20,  20,  20,  20,  20,  20,  -4,  -4,  -4,  20,  -4,  -4,
    -4,  -4,  -4,  20,  -4,  -4,  -4,  -4,  -4,  -4,  -4,  20,  -4,  20,  20,
    22,  -4,  -4,  -4,  -4,  -4,  -4,  -4,  -4,  -4,  -4,  -4,  -4,  23,  23,
    -4,  23,  23,  -4,  -4,  -4,  -4,  23,  -4,  -4,  -4,  -4,  -4,  -4,  -4,
    -4,  -4,  -4,  -4,  -4,  -4,  -4,  24,  24,  -4,  24,  24,  -4,  -4,  -4,
    -4,  24,  -4,  -4,  -4,  -4,  -4,  -4,  -4,  -4,  -4,  -4,  -4,  -4,  -4,
    -4,  25,  25,  -4,  25,  25,  -4,  -4,  -4,  -4,  25,  -4,  -4,  -4,  -4,
    -4,  -4,  -4,  -4,  -4,  -4,  -4,  -4,  -4,  -4,  20,  20,  -4,  20,  20,
    -4,  -4,  -4,  -4,  20,  -4,  -4,  -4,  -4,  -13, -13, -13, -13, -13, -13,
    -13, -13, -13, -13, -13, -13, -13, -13, -13, -13, -13, -13, -13, -13, -13,
    -13, -13, -13, -11, -11, 27,  -11, -11, -11, -11, -11, -11, -11, -11, -11,
    -11, -11, -11, -11, -11, -11, -11, -11, -11, -11, -11, -11,
};

constexpr int tmBacktracking[] = {
    11, 11,  // in JSONNumber
    11, 8,   // in JSONNumber
};
}  // namespace

Token Lexer::Next() {
restart:
  token_line_ = line_;
  token_offset_ = offset_;

  int state = tmStateMap[start_state_];
  uint32_t hash = 0;
  Token backupToken = Token::EOI;
  uint64_t backupOffset;
  uint32_t backupHash = hash;

  while (state >= 0) {
    int curr_class;
    if (input_rune_ < 0) {
      state = tmLexerAction[state * tmNumClasses];
      if (state > tmFirstRule && state < 0) {
        state = (-1 - state) * 2;
        backupToken = Token(tmBacktracking[state]);
        backupOffset = offset_;
        backupHash = hash;
        state = tmBacktracking[state + 1];
      }
      continue;
    } else if (input_rune_ < tmRuneClassLen) {
      curr_class = tmRuneClass[input_rune_];
    } else {
      curr_class = 1;
    }
    state = tmLexerAction[state * tmNumClasses + curr_class];
    if (state > tmFirstRule) {
      if (state < 0) {
        state = (-1 - state) * 2;
        backupToken = Token(tmBacktracking[state]);
        backupOffset = offset_;
        backupHash = hash;
        state = tmBacktracking[state + 1];
      }
      hash = hash * 31 + static_cast<uint32_t>(input_rune_);

      if (input_rune_ == '\n') {
        line_++;
      }

      // Scan the next character.
      // Note: the following code is inlined to avoid performance
      // implications.
      offset_ = scan_offset_;
      if (offset_ < source_.size()) {
        // UTF8 is not supported.
        input_rune_ = static_cast<uint8_t>(source_[scan_offset_++]);
      } else {
        input_rune_ = -1;
      }
    }
  }

  Token tok = Token(tmFirstRule - state);
recovered:
  switch (tok) {
    case Token::ID: {
      uint32_t hh = hash & 7;
      switch (hh) {
        case 1:
          if (hash == 0x41 && 'A' == source_[token_offset_]) {
            tok = Token::CHAR_A;
          }
          break;
        case 2:
          if (hash == 0x42 && 'B' == source_[token_offset_]) {
            tok = Token::CHAR_B;
          }
          break;
        case 3:
          if (hash == 0x5cb1923 && "false" == Text()) {
            tok = Token::FALSE;
          }
          break;
        case 6:
          if (hash == 0x36758e && "true" == Text()) {
            tok = Token::TRUE;
          }
          break;
        case 7:
          if (hash == 0x33c587 && "null" == Text()) {
            tok = Token::JSON_NULL;
          }
          break;
      }
      break;
    }
    default:
      break;
  }
  switch (tok) {
    case Token::INVALID_TOKEN:
      if (static_cast<int>(backupToken) >= 0) {
        tok = Token(backupToken);
        hash = backupHash;
        Rewind(backupOffset);
      } else if (offset_ == token_offset_) {
        Rewind(scan_offset_);
      }
      if (tok != Token::INVALID_TOKEN) {
        goto recovered;
      }
      break;
    case Token(8):
      goto restart;
    default:
      break;
  }
  VLOG(2) << "Lexer::Next returning next token " << tok << " now " << *this;
  return tok;
}

Lexer::Lexer(absl::string_view input_source) {
  source_ = input_source;
  if (absl::StartsWith(source_, bomSeq)) {
    offset_ += bomSeq.size();
  }
  Rewind(offset_);
}

void Lexer::Rewind(int64_t rewind_offset) {
  if (rewind_offset < offset_) {
    for (int64_t i = rewind_offset; i < offset_; ++i) {
      if (source_[i] == '\n') {
        line_--;
      }
    }
  } else {
    if (rewind_offset > source_.size()) {
      rewind_offset = source_.size();
    }
    for (int64_t i = offset_; i < rewind_offset; ++i) {
      if (source_[i] == '\n') {
        line_++;
      }
    }
  }

  // Scan the next character.
  scan_offset_ = rewind_offset;
  offset_ = rewind_offset;
  if (offset_ < source_.size()) {
    input_rune_ = source_[scan_offset_++];
  } else {
    input_rune_ = -1;  // Invalid rune for end of input
  }
}
}  // namespace json
